Disallow overriding `==` and `hashCode`. (#2118)
dart-lang/language: accepted/future-releases/enhanced-enums/feature-specification.md

where `memberDeclaration*` is almost any sequence of static and instance member declarations, or constructors, 
with some necessary restrictions specified below.

> The `;` after the value list is optional if there is nothing else in the declaration (for backwards compatibility), and required if there is any member declaration after it. The identifier list may have a trailing comma in either case (like now).

Reviewer: Shouldn't this be also changed to `value`?

```suggestion
The `;` after the value list is optional if there is nothing else in the declaration (for backwards compatibility), and required if there is any member declaration after it. The value list may have a trailing comma in either case (like now).
```

  It’s a **compile-time error** if such a mixin application introduces any instance variables. _We need to be able to call an implementation specific superclass `const` constructor of `Enum`, and a mixin application of a mixin with a field does not make its forwarding constructor `const`. Currently that’s the only restriction, but if we add further restrictions on mixin applications having `const` forwarding constructors, those should also apply here._

> * **Superinterfaces**: The immediate superinterfaces of *C* are the interface of the superclass and the interfaces declared bu `implements` in *E*.

Reviewer 2: ```suggestion
* **Superinterfaces**: The immediate superinterfaces of *C* are the interface of the superclass and the interfaces declared by `implements` in *E*.
```

  where `args` are considered as occurring in a `const` context, and it’s a **compile-time error** if they are then not compile-time constants.

>   The resulting constructor invocations are subject to type inference, using the empty context type. *This implies that inferred type arguments to the constructor invocation itself may depend on the types of the argument expressions of `args`.* The type of the constant variable is the static type of the resulting constant object creation expression.

Author: ... is **worst** kind of whitespace!

   The objects created here are *not canonicalized* like other constant object creations. _(In practice, the index value is considered part of the object, so no two objects will have the same state.)_

- **Static `values` list**: A static constant variable named `values` is added as by the declaration  `static const List<Name> values = [id1, …, idn];`
  where `id1`…`idn` are the names of the enum entries of the `enum` declaration in source/index order.
  _If `Name` is generic, the `List<Name>` instantiates `Name` to its bounds._

> It's a **compile-time error** if an `enum` declaration declares or inherits (from a mixin application) a non-abstract member named  `index` which overrides the `index` getter inherited from the `Enum` class.

Author: Agree. I've moved the mention of the mixin application to after the paragraph, and comment-italicized it.

> It's a **compile-time error** if an `enum` declaration declares or inherits (from a mixin application) a non-abstract member named  `hashCode` or `==` (an `operator ==` declaration) which overrides the `hashCode` getter or `==` operator inherited from the `Object` class. (The `Enum` class does not override `hashCode` or `operator==` from `Object`). _This ensures that enum values can be used as switch statement case values, which is the main advantage of using an enum over just writing a normal class._

Author: I'm also vary about using "has".

A declaration *contains* other declarations. It's actually even a stretch to say that a declaration *inherits* anything.
The semantic class/mixin/enum that the declaration introduces may have thing or inherit things, but if there is a compile-time error, then that class/mixin/enum does not exist, so it can't "have" anything.

So, to define things in terms of concepts that are well-defined at this point in the validation, I refer only to member declarations of the member itself, or the class/interface of its superclass/superinterfaces (which are assumed well-formed before we even start looking at this declaration). The "inherits" here is really a shorthand for "the superclass has X so the enum would inherit X, if it is otherwise valid".

> It's a **compile-time error** if an `enum` declaration declares or inherits (from a mixin application) a non-abstract member named  `hashCode` or `==` (an `operator ==` declaration) which overrides the `hashCode` getter or `==` operator inherited from the `Object` class. (The `Enum` class does not override `hashCode` or `operator==` from `Object`). _This ensures that enum values can be used as switch statement case values, which is the main advantage of using an enum over just writing a normal class._

Author: Although in this case, it's probably workable, since if the enum declaration is otherwise valid, we can *then* check if it has an `index`/`hashCode`/`==` instance member which overrides the ones from `Enum`/`Object`.

I still don't like "has". It's too vague.

- It is a **compile-time error** if a class implements, extends or mixes-in the class or interface introduced by an `enum` declaration. _(An enum class can’t be used as a mixin since it is not a `mixin` declaration and the class has a superclass other than `Object`, but we include “mixes-in” for completeness.)_
- It's a **compile-time error** if a `class` or `mixin` declaration has `Enum` as a superinterface and the interface of the declarations contains an instance member with the name `values`, whether declared or inherited. _If any concrete class implements this interface, it will be an `enum` declaration class, and then the `values` member would conflict with the static `values` constant getter that is automatically added to `enum` declaration classes. Such an instance `values` declaration is either useless or wrong, so we disallow it entirely._

> - It's a **compile-time error** if a `class` or `mixin` declaration has `Enum` as a superinterface, and that class or mixin declares or inherits a concrete instance member named `index`, `hashCode` or `==` *(an `operator ==` declaration)*. _That `index` member could override the `index` getter inherited from `Enum`, and we currently do not allow that. The `hashCode` and `operator==` declarations would prevent the enum class from having "primitive equality", and we want to ensure that enums can be used in switches._

Reviewer 3: A mixin actually cannot inherit a concrete instance member, because the superclass in any mixin application is only required to implement any `on` types of the mixin, an `on` type doesn't have to be a superclass.

But I guess that's benign, because it just means that the 'mixin' case "can't use the word 'inherits'", it doesn't introduce anything which is wrong.

- It is a compile-time error if a class implements, extends or mixes-in the class or interface introduced by an `enum` declaration. _(An enum class can’t be used as a mixin since it is not a `mixin` declaration and the class has a superclass other than `Object`, but we include “mixes-in” for completeness.)_
- It's a **compile-time error** if a `class` or `mixin` declaration has `Enum` as a superinterface and the interface of the declarations contains an instance member with the name `values`, whether declared or inherited. _If any concrete class implements this interface, it will be an `enum` declaration class, and then the `values` member would conflict with the static `values` constant getter that is automatically added to `enum` declaration classes. Such an instance `values` declaration is either useless or wrong, so we disallow it entirely._

> - It's a compile-time error if a `class`, `mixin` or `enum` declaration has `Enum` as a superinterface (trivially true for `enum` declarations), and that declaration contains non-abstract instance member declaration named `index`, `hashCode` or `==` (an `operator ==` declaration). _That `index` member would override the `index` getter inherited from `Enum`, and we currently do not allow that. The `hashCode` and `operator==` declarations would prevent the enum class from having "primitive equality", and we want to ensure that enums can be used in switches._

Reviewer 3: A couple of changes would fit here, for consistency:
```suggestion
- It's a compile-time error if a `class`, `mixin` or `enum` declaration has `Enum` as a superinterface (*trivially true for `enum` declarations*), and that declaration contains a concrete instance member declaration named `index`, `hashCode` or operator `==`. _That `index` member would override the `index` getter inherited from `Enum`, and we currently do not allow that. The `hashCode` and operator `==` declarations would prevent the enum class from having "primitive equality", and we want to ensure that enums can be used in switches._
```

However, there is another issue: This item repeats rules that have already been stated (in line 134 and 136), and this normally implies that the duplicate should be commentary. A similar issue arises for `values` in line 160.

Author: This actually *doesn't* duplicate the earlier errors.

This error only apply to *non-abstract declarations* of `index`/`hashCode`/`==` in a `class`/`mixin`/`enum` which implements `Enum`.

The errors in lines 134/136 apply to an `enum` declaration which declares or inherits `index`/`hashCode`/`==`.

So, the case of an `enum` with a mixin application of a mixin *not* implementing `Enum`, but declaring a non-abstract `index`/`hashCode`/`==`, is only covered by the earlier errors, not this one.

And that's probably an oversight, because we don't cover the case of:
```dart
class C {
  int get index => 0;
}
abstract class D extends C implements Enum {}
```
either.

It's actually not particularly *important* that a `class` implementing `Enum` has a concrete implementation of anything, because there is currently no way to ever get that implementation into an `enum` declaration.
(This error catches the class itself declaring `index`, etc., and if it's declared in a superclass, then the class has a superclass other than `Objecf` and cannot be used as a mixin.)

So, how strict should we be about classes implementing `Enum` and having implementations. Those implementations are guaranteed dead code, so we could choose to refuse it entirely.
Or we could just not care (status quo), since you'll probably figure it out if you ever try to use the class for anything except `implements` its interface.

Author: I actually prefer "non-abstract" to "concrete" because being `abstract` is the actual language term.

Reviewer 3: The language specification has 58 occurrences of `concrete` and zero occurrences of 'non-abstract', so we do have prior art to support the notion that 'concrete' is well-defined. We could switch over, of course, but 'non-abstract' does seem somewhat bulky and hard to pronounce.

Author: The
> instance member declaration named `index`, `hashCode` or operator `==`

phrase doesn't work for me. There is no "instance member declaration named ... operator `==`".
It's either *named* `==` or it's an `operator ==` declaration.

Reviewer 3: NB: I edited several 'suggestions' above to fit the results of the discussion at this point. Added `[Edit: ...]` there.

> The errors in lines 134/136 apply to an `enum` declaration

Indeed, I forgot about that distinction even though it came up earlier in discussions with the UniPro folks. ;-)

I changed the edit suggestion such that the overlap is reduced: The rule here will then apply to classes and mixins, not to enum declarations.

However, these error specifications should occur right next to each other, it is too hard to read the document when they are so far apart from each other.

> So, the case of an `enum` with a mixin application of a mixin not implementing `Enum`, but declaring a non-abstract `index`/`hashCode`/`==`, is only covered by the earlier errors, not this one.

Right, and we do have cases where line 161 applies and line 134/136 do not:

```dart
mixin M on Enum {
  int get index => 0; // Error.
}

abstract class B implements Enum {
  int get index => 0; // Error.
}

class C {
  int get index => 0;
}
abstract class D extends C implements Enum {} // Error.
```

I suggested that we should use the phrase '.. C has a concrete instance member `m`..', meaning that the superclass chain of `C` contains an element `B` (which could be `C` or some other class, possibly a mixin application) such that `B` declares a concrete instance member named `m`.

There is a bit of a conceptual clash here, because 'has'/'declares' is usually used with interfaces, that is, with member signatures and not actual declarations, but I think it is quite readable and not ambiguous in practice.

> Those restrictions allow abstract classes (interfaces) which *implements* `Enum` in order to have the `int index;` getter member available, and also allow `mixin` declarations to use `Enum` as an `on` type because `mixin` declarations cannot be instantiated directly.

Reviewer 4: implements -> implement

Reviewer 3: The language specification uses `\IMPLEMENTS` as a reference to the actual syntax and 'implements' as a separate concept (such that `List<int>` implements `Iterable<int>` and `Object`, but not `Iterable<num>`). In this context I think we could use 'abstract classes that implement `Enum`' and remain consistent with the way this word is used in the spec.

We could include the word 'interface', but I think that's more opinion than information. ;-)

> The restrictions still ensure  `enum` values are the only object instances which implements `Enum`, while making it valid to declare `abstract class MyInterface implements Enum` and `mixin MyMixin on Enum` for interfaces and mixins intended to be used in declaring `enum` classes. It's also impossible to override the instance getters `index` , `hashCode`, or the operator `==`, and impossible to interfere with the static  `values` getter, without causing a compile-time error. _For example, if an enum declaration implements an interface containing the signature `Never get index;`, then because it's not possible to override `int get index;` from `Enum`, the resulting class will not implement its interface and that is a compile-time error._

Reviewer 3: The paragraph as a whole seems to be commentary in nature.
```suggestion
_The restrictions still ensure that `enum` values are the only objects whose run-time type implements `Enum`, while making it valid to declare `abstract class MyInterface implements Enum` and `mixin MyMixin on Enum` for interfaces and mixins intended to be used in `enum` declarations. It's also impossible to override the instance getters `index` , `hashCode`, or the operator `==`, and impossible to interfere with the static  `values` getter, without causing a compile-time error. For example, if an enum declaration implements an interface containing the signature `Never get index;`, then because it's not possible to override `int get index;` from `Enum`, the resulting class will not implement its interface and that is a compile-time error._
```

Enum instances are objects like any other object, and with this change they can implement interfaces and inherit members from mixins. The main difference between an `enum` declaration and a hand-written “equivalent class” using the enum pattern is that:

- The `enum` types implement `Enum`. The `Enum` type is otherwise sealed against instantiation, so no other *objects* than enum entries can implement it. Abstract classes and mixins can implement or extend `Enum`, but unless they are then implemented by or mixed into an `enum` declaration, no objects can actually implement the type of that class or mixin.
- The `enum` types themselves are completely sealed. No other class can implement an `enum` type.

- Because of that, `enum` types support exhaustiveness checking in `switch` cases in the language _(meaning that flow-control can see that an exhaustive switch over enum values cannot pass through without executing at least one `case`, which can then affect variable promotion)_.
- The `EnumName.name` extension member works on `enum` values.

> If the *restrictions* (the type is sealed, there is only a finite, enumerable number of instances, and the class  implements `Enum`, so it must have an `int index` getter and no `values` member, and it cannot override `index`, `hashCode` or `operator==`), are acceptable, there should no longer be any reason to *not* make your enum-like class a language-based `enum`.

Reviewer 5: > ...so it must have an `int index` getter and no `values` member, and it cannot override `index`...

Must have `int index` and cannot override `index` sounds conflicting.

Reviewer 3: I think we can just avoid mentioning that `index` must be there (that's implied by the following 'cannot override `index`' anyway). With a couple of other adjustments:

```suggestion
If the *restrictions* are acceptable (the type is sealed; there is only a finite, statically known set of instances; the class  implements `Enum`; it cannot override `index`, `hashCode` or operator `==`; and it cannot have a user-written member named `values`), there should no longer be any reason to *not* make your enum-like class a language-based `enum`.
```
